File: python/dvforge/compiler.py
from __future__ import annotations

import shutil
from pathlib import Path

from dvforge.generators import (
    attribute,
    entity,
    formxml,
    optionset,
    publisher,
    relationship,
    ribbondiff,
    savedquery,
    solution,
)
from dvforge.model import Config
from dvforge.utils import write_yaml


def compile(config: Config, output_dir: Path, managed: bool = True) -> None:
    prefix = config.solution.publisher.prefix
    files: dict[str, dict] = {}

    # Publisher
    files.update(publisher.generate(config.solution.publisher))

    # Option sets
    for os in config.option_sets:
        files.update(optionset.generate(os, prefix))

    # Per-entity generators
    for ent in config.entities:
        files.update(entity.generate(ent, prefix))
        files.update(attribute.generate(ent, prefix))
        files.update(formxml.generate(ent, prefix))
        files.update(savedquery.generate(ent, prefix))
        files.update(ribbondiff.generate(ent, prefix))
        files.update(relationship.generate(ent, prefix))

    # Solution files last — needs all component paths
    files.update(solution.generate(config, managed, list(files.keys())))

    # Clean output directory then write all files
    if output_dir.exists():
        shutil.rmtree(output_dir)
    output_dir.mkdir(parents=True)

    for rel_path, data in files.items():
        write_yaml(output_dir / rel_path, data)

File: npx/compiler.ts
import fs from "fs";
import path from "path";
import { Config } from "./model.js";
import { writeYaml } from "./utils.js";
import * as publisher from "./generators/publisher.js";
import * as optionset from "./generators/optionset.js";
import * as entity from "./generators/entity.js";
import * as attribute from "./generators/attribute.js";
import * as formxml from "./generators/formxml.js";
import * as savedquery from "./generators/savedquery.js";
import * as ribbondiff from "./generators/ribbondiff.js";
import * as relationship from "./generators/relationship.js";
import * as solution from "./generators/solution.js";

export function compile(config: Config, outputDir: string, managed = true): void {
  const publisherPrefix = config.solution.publisher.prefix;
  const files = new Map<string, unknown>();

  // publisher
  for (const [k, v] of Object.entries(publisher.generate(config.solution.publisher)))
    files.set(k, v);

  // option sets
  for (const os of config.option_sets)
    for (const [k, v] of Object.entries(optionset.generate(os, publisherPrefix)))
      files.set(k, v);

  // entity
  for (const ent of config.entities) {
    for (const gen of [entity, attribute, formxml, savedquery, ribbondiff, relationship])
      for (const [k, v] of Object.entries(gen.generate(ent, publisherPrefix)))
        files.set(k, v);
  }

  // solution
  for (const [k, v] of Object.entries(solution.generate(config, managed, [...files.keys()])))
    files.set(k, v);

  // clean output dir, write all files
  fs.rmSync(outputDir, { recursive: true, force: true });
  fs.mkdirSync(outputDir, { recursive: true });
  for (const [relPath, data] of files)
    writeYaml(path.join(outputDir, relPath), data);
}
